feat: Introduce Reverification (#1726)
clerk/clerk-docs: docs/guides/re-verification.mdx

---
title: Get started with Re-verification
description: Leverage Clerk’s utilities, to frequently ask users to verify their identity before performing sensitive actions.
---

## Introduction

> After a user signs into your application, you may want to control their access and actions based on how long they have been logged in.

Author: Maybe I need to rephrase, It's based on the time the last verified their credentials, but logging in, counts as a verification.

For example, in a banking application, transferring money between accounts is considered a “sensitive action.” In such cases, you may require the user to “re-verify” their credentials before completing the action.

Clerk provides a suite of utilities to help you implement a seamless re-verification flow.

- `auth().has()` now accepts a `reverification` attribute

> - `useReverification()` automatically displays a modal prompting the end-user to reverify their credentials and resume the sensitive action.

Author: I've updated the text

    ```tsx {{ filename: '/app/transfer/page.tsx' }}
    'use client'

    import { useReverification } from '@clerk/nextjs'

    export function TransferButton({ amount_in_cents }) {
      const [transferMoney] = useReverification(() =>
        fetch('/api/transfer-balance', {
          method: 'POST',
          body: JSON.stringify({ amount_in_cents }),
        }),
      )

      return <button onClick={transferMoney}>Transfer</button>
    }
    ```
  </Tab>

  <Tab>
    In your controller add a filter that leverages the new helpers

    ```ruby
    # app/controllers/home_controller.rb
    class HomeController < ApplicationController
      before_action :require_reverification, only: :foo

      def foo
        render json: { sucess: "true" }
      end

      private
      
      # will halt the request and respond with a JSON that Clerk.js

>       # will read and kickstart a re-verification flow 

Reviewer: ```suggestion
      # will read and kickstart a reverification flow 
```

    In your react application you can simply wrap your fetcher with `useReverification()`.

    ```tsx
    import { useReverification } from '@clerk/react'

    export function TransferButton({ amount_in_cents }) {
      const [transferMoney] = useReverification(() =>
        fetch('/api/transfer-balance', {
          method: 'POST',
          body: JSON.stringify({ amount_in_cents }),
        }),
      )

      return <button onClick={transferMoney}>Transfer</button>
    }
    ```
  </Tab>

  <Tab>
    You can use the `reverificationMismatchResponse` exported from `@clerk/shared/authorization-errors` . If you are not using Javascript / Typescript, you can create your own helper that returns this JSON response:

    ```json
    {
      "clerk_error": {
        "type": "forbidden",
        "reason": "reverification-mismatch"
      }
    }
    ```

    It is also advised to use a 403 status code on your response.
  </Tab>
</Tabs>

## Caveats

Before opting-in to this feature, there are a few important considerations:

1. **Available Factors for Reverification**: Not all authentication factors can be used for reverification.

The available options are:

> - First factors: Password, Email code, Phone code

Reviewer: ```suggestion
- First factors: password, email code, phone code
```

> - Second Factors: Phone code, Authenticator app, Backup code

Reviewer: ```suggestion
- Second factors: phone code, authenticator app, backup code
```

> 1. **Eligibility for Sensitive Actions**: If a user does not have any of the above factors configured, they will be unable to reverify their credentials and therefore ineligible to perform “sensitive actions.”

Reviewer: ```suggestion
1. **Eligibility for Sensitive Actions**: If a user does not have any of the above factors available, they will be unable to reverify. This is a concern for applications that do not require email addresses to sign up, and applications that have disabled email codes in favor of email links.”
```

> 1. **Graceful Downgrade of Verification Level**: If you request a `second_factor` or `multi_factor` level of verification but the user does not have the required factor available, our utilities automatically downgrade the requested level to `first_factor` to ensure a smoother experience.

Reviewer: ```suggestion
1. **Graceful Downgrade of Verification Level**: If you request a `second_factor` or `multi_factor` level of verification but the user does not have a second factor available, our utilities automatically downgrade the requested level to `first_factor`.
```

Reviewer: Also, move this above **Eligibility for Sensitive Actions**

## Wrap up

> Re-verification helps secure sensitive actions by ensuring users periodically confirm their identity. With Clerk’s tools like `auth().has()` and `useReverification()`, you can seamlessly implement this feature while maintaining a smooth user experience.

Reviewer: ```suggestion
```

        clerk_render_reverification if clerk_user_needs_reverification?
      end
    end
    ```

    ### Handling authorization errors on the client

    In your react application you can simply wrap your fetcher with `useReverification()`.

    ```tsx
    import { useReverification } from '@clerk/react'

    export function TransferButton({ amount_in_cents }) {
      const [transferMoney] = useReverification(() =>
        fetch('/api/transfer-balance', {
          method: 'POST',
          body: JSON.stringify({ amount_in_cents }),
        }),
      )

      return <button onClick={transferMoney}>Transfer</button>
    }
    ```
  </Tab>

  <Tab>
    You can use the `reverificationMismatchResponse` exported from `@clerk/shared/authorization-errors` . If you are not using Javascript / Typescript, you can create your own helper that returns this JSON response:

    ```json
    {
      "clerk_error": {
        "type": "forbidden",
        "reason": "reverification-mismatch"
      }
    }
    ```

    It is also advised to use a 403 status code on your response.
  </Tab>
</Tabs>

## Caveats

Before opting-in to this feature, there are a few important considerations:

> 1. Available Factors for Reverification
>   Not all authentication factors can be used for reverification. The available options are:

Author: ```suggestion
1. **Available Factors for Reverification**: Not all authentication factors can be used for reverification.

The available options are:
```

- First factors: Password, Email code, Phone code
- Second Factors: Phone code, Authenticator app, Backup code

> 1. Eligibility for Sensitive Actions

Author: ```suggestion
1. **Eligibility for Sensitive Actions**:
```

> 1. Eligibility for Sensitive Actions
>   If a user does not have any of the above factors configured, they will be unable to reverify their credentials and therefore ineligible to perform “sensitive actions.”

Author: ```suggestion
1. **Eligibility for Sensitive Actions**: If a user does not have any of the above factors configured, they will be unable to reverify their credentials and therefore ineligible to perform “sensitive actions.”
```

> 1. Graceful Downgrade of Verification Level
>   If you request a `second_factor` or `multi_factor` level of verification but the user does not have the required factor available, our utilities automatically downgrade the requested level to `first_factor` to ensure a smoother experience.

Author: ```suggestion
1. **Graceful Downgrade of Verification Level**: If you request a `second_factor` or `multi_factor` level of verification but the user does not have the required factor available, our utilities automatically downgrade the requested level to `first_factor` to ensure a smoother experience.
```
